[WIP] Porting Masthead to Vue.js

inkuzmin opened the pull request:
This is a work in progress on porting the masthead to Vue.js.

- [x] Add the new masthead below the original one
- [x] Achieve markup parity using https://bootstrap-vue.js.org/docs/components/navbar/ (thanks @guerler for the hint)
- [x] Add tooltips
- [x] "Please login or register to use this feature."
- [x] Add actions (links/routes), skip the scratch-book for a moment
- [x] Figure out what is the "nav-note" and port it
- [x] Figure out what the "noscratchbook" property is responsible for
- [x] Add the scratch-book
- [x] Highlight an active item correctly
- [x] Use dot notation for properties accessors
- [x] Refactor NavItem into a separate component
- [x] Handle dividers at the dropdown menu
- [x] Refactor NavItem classes as computed properties
- [x] Remove the original NavBar
- [ ] Check if any functionality is still missing and if that's the case port it
- [ ] Remove covered code
- [ ] Prettify code, make tests pass

inkuzmin committed:
port masthead to vue

inkuzmin committed:
add tooltips

inkuzmin committed:
add popovers

inkuzmin committed:
actions

inkuzmin committed:
port nav-note

inkuzmin committed:
add scratch book

inkuzmin committed:
active tab

inkuzmin committed:
fix scratch-book indicator

guerler commented on client/galaxy/scripts/components/Masthead.vue:
```diff
@@ -0,0 +1,210 @@
+                                            tab.icon || ''
+                                        ]">
+
+                                {{ tab["title"] }}
```
Please use `tab.title` here and at other similar locations.

guerler commented on client/galaxy/scripts/components/Masthead.vue:
```diff
@@ -0,0 +1,210 @@
+
+                <b-navbar-nav>
+                    <template v-for="tab in tabs">
+                            <b-nav-item v-if="!tab.menu"
```
This `b-nav-item` (maybe together with `b-nav-item-dropdown`) should be in its own subcomponent e.g. MastheadItem.vue, then the below class and property settings can be computed values. It will simplify and improve readability of styling and configuring an individual nav-item. Other than that this looks great, thank you so much for working on this.

inkuzmin committed:
dot notation for property accessors

inkuzmin committed:
prepare nav item to be refactored into a separate component

inkuzmin committed:
refactor nav item (wip)

inkuzmin committed:
refactor some attributes as computed properties

inkuzmin committed:
format

inkuzmin committed:
remove the original navbar

guerler committed:
Merge branch 'dev' into navbar-vue

guerler committed:
Move masthead components to masthead directory

guerler committed:
Format client

guerler committed:
Remove unused template helper

guerler commented:
Thank you so much for working on this. I noticed that there is an issue with the navigation, an error is triggered when navigating to the admin panel and then back to the workflow list. I also noticed that the masthead is not rendered for the data libraries view.

inkuzmin commented:
@guerler, thank you for checking it! I am going to try to fix the issues you've noticed tomorrow.

inkuzmin committed:
fix multilevel links

inkuzmin committed:
show navbar at the libraries page

guerler commented on client/galaxy/scripts/layout/masthead.js:
```diff
@@ -81,34 +71,38 @@ const View = Backbone.View.extend({
+                activeTab: () => {
+                    return this.activeView;
+                },
+                tabs: _.map(this.collection.models, el => {
```
I recommend to use built-in map `array.map()` instead of underscore.

guerler commented on client/galaxy/scripts/layout/masthead.js:
```diff
@@ -81,34 +71,38 @@ const View = Backbone.View.extend({
+                brandLink: this.options.logo_url,
+                brandImage: this.options.logo_src,
+                quotaMeter: this.quotaMeter,
+                activeTab: () => {
```
Here we could parse `this.activeView` directly instead of wrappig it into a function.

guerler committed:
Parse tab classes, fix failing anon selenium test case, format client

guerler committed:
Merge branch 'dev' into navbar-vue

guerler committed:
Restore swapped cls attribute

guerler commented on client/galaxy/scripts/components/Masthead/Masthead.vue:
```diff
@@ -0,0 +1,117 @@
+<template>
+    <div>
+        <b-navbar
```
This wrapped twice, you can remove the outer div-tag.

jmchilton committed:
Merge remote-tracking branch 'galaxyproject/dev' into navbar-vue

jmchilton committed:
Reformat and lint fixes.

jmchilton committed:
Eliminate extra wrapper div in Masthead vue component.

jmchilton committed:
Integrate Sam's suggestions on masthead PR.

- Switch activeTab to a String instead of a Function.
- Use .map instead of underscore (eliminate import).

jmchilton committed:
More computed properties in MastheadItem.

jmchilton committed:
Fix masthead sometimes being too tall.

jmchilton committed:
Merge remote-tracking branch 'galaxyproject/dev' into navbar-vue

jmchilton committed:
Add simple, minimal unit tests for Masthead*.vue.

jmchilton committed:
More masthead unit tests.

Implement hidden tabs also to match unit tests in qunit.

jmchilton committed:
Masthead - add unit test for active tab.

jmchilton committed:
More progress on masthead tests.

jmchilton committed:
Remove now unused menu tab code in layout/menu.js.

jmchilton committed:
Fix Selenium tests for masthead change.

jmchilton committed:
Another selenium fix.

jmchilton committed:
Another Selenium fix for masthead changes.

jmchilton committed:
More selenium test fixes for new masthead.

Problem is the interactive tooltips that are off by default in bootstrap but not bootstrap-vue. interactive tooltips are better for accessibility though so best to just work around them for testing?

jmchilton committed:
Remove masthead qunit tests (most ported to VueJS tests)

dannon commented on client/galaxy/scripts/components/Masthead/MastheadItem.vue:
```diff
@@ -0,0 +1,182 @@
+        v-b-popover.manual.bottom="{ id: tab.id, content: popoverNote, html: true }"
+        @click="open(tab, $event)"
+    >
+        <template v-if="tab.icon">
```
It'd be great if future work could keep in mind and use the new(ish) vue-fontawesome work that was introduced to swap out the old style fontawesome classes and includes.

dannon commented on lib/galaxy/selenium/navigation.yml:
```diff
@@ -31,21 +31,19 @@ masthead:
-    user:
-      type: xpath
-      selector: '//li[@id="user"]'
+    # bootstrap-vue a tag doesn't work as link target, need to hit span inside
```
Hopefully this doesn't break many tours.
